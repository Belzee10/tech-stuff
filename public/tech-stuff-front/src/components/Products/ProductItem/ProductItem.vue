<template>
  <v-card class="product-item" @click.stop="cardClick">
    <v-container class="py-0">
      <v-row justify="space-between">
        <v-col class="pa-0" cols="auto">
          <v-img
            height="100"
            width="100"
            :src="image"
            lazy-src="https://picsum.photos/id/11/10/6"
          ></v-img>
        </v-col>

        <v-col class="pa-0" cols="5">
          <div class="py-2 d-flex flex-column fill-height">
            <h4 class="body-1">{{ name }}</h4>
            <span class="caption">{{ brand }}</span>
            <span class="font-weight-bold price d-block mt-auto"
              >${{ price }}</span
            >
          </div>
        </v-col>

        <v-col cols="2" class="text-center pa-0">
          <v-row class="flex-column ma-0 fill-height justify-space-between">
            <v-col
              v-if="inStock > 0"
              class="pa-0 d-flex flex-column justify-center align-center"
            >
              <v-btn
                icon
                color="secondary"
                class="btn-small add"
                @click.stop="addToCart"
              >
                <v-icon small>mdi-cart-plus</v-icon>
              </v-btn>
            </v-col>
            <v-col
              v-if="inStock > 0"
              class="pa-0 d-flex flex-column justify-center align-center"
            >
              <span>{{ inStock }}</span>
            </v-col>
            <v-col
              v-if="inStock > 0"
              class="pa-0 d-flex flex-column justify-center align-center"
            >
              <v-btn
                icon
                color="secondary"
                class="btn-small remove"
                @click.stop="removeFromCart"
              >
                <v-icon small>mdi-cart-minus</v-icon>
              </v-btn>
            </v-col>
            <v-col
              v-else
              class="pa-0 d-flex flex-column justify-center align-center"
            >
              <v-btn
                icon
                color="secondary"
                class="btn-full"
                @click.stop="addToCart"
              >
                <v-icon small>mdi-cart-plus</v-icon>
              </v-btn>
            </v-col>
          </v-row>
        </v-col>
      </v-row>
    </v-container>
  </v-card>
</template>

<script>
import { productMixin } from '@/mixins';

export default {
  name: 'ProductItem',
  mixins: [productMixin]
};
</script>

<style lang="scss" scoped></style>
