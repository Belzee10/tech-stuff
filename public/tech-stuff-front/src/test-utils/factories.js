import { mount, createLocalVue } from '@vue/test-utils';
import mergeWith from 'lodash.mergewith';
import Vuetify from 'vuetify';
import Vuex from 'vuex';

const localVue = createLocalVue();
localVue.use(Vuex);

let vuetify;

beforeEach(() => {
  vuetify = new Vuetify();
});

const customizer = (_, srcValue) => {
  if (Array.isArray(srcValue)) {
    return srcValue;
  }
  if (srcValue instanceof Object && Object.keys(srcValue).length === 0) {
    return srcValue;
  }
};

/**
 * create component wrapper
 * @param {*} Component
 * @param {Object} overrides
 */
const createWrapper = (Component, overrides) => {
  const defaultMountingOptions = {
    localVue,
    vuetify
  };
  return mount(
    Component,
    mergeWith(defaultMountingOptions, overrides, customizer)
  );
};

/**
 * create Vuex store
 * @param {Object} overrides
 */
const createStore = overrides => {
  const defaultStoreConfig = {};
  return new Vuex.Store(mergeWith(defaultStoreConfig, overrides, customizer));
};

export { createWrapper, createStore };
