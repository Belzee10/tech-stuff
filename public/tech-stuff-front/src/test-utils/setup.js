import Vue from 'vue';
import Vuetify from 'vuetify';

Vue.config.productionTip = false;

Vue.use(Vuetify);

/**
 * silent "console.error"
 */
global.console = {
  error: jest.fn(),
  log: console.log,
  warn: console.warn,
  info: console.info,
  debug: console.debug
};
