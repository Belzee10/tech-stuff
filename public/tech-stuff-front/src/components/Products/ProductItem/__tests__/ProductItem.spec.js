import { createWrapper } from '@/test-utils/factories';
import ProductItem from '../ProductItem.vue';

describe('ProductItem.vue', () => {
  test('should emit a "add-to-cart" event', () => {
    const props = {
      id: 1
    };
    const wrapper = createWrapper(ProductItem, {
      propsData: props
    });
    wrapper.find('.v-btn').trigger('click');
    expect(wrapper.emitted('add-to-cart')[0]).toEqual([props.id]);
  });
  test('should render two "btn-small", emit "add-to-cart" and "remove-from-cart" events', () => {
    const props = {
      id: 1,
      inStock: 2
    };
    const wrapper = createWrapper(ProductItem, {
      propsData: props
    });
    expect(wrapper.find('.btn-full').exists()).toBeFalsy();
    expect(wrapper.findAll('.btn-small')).toHaveLength(2);
    wrapper.find('.btn-small.remove').trigger('click');
    expect(wrapper.emitted('remove-from-cart')[0]).toEqual([props.id]);
    wrapper.find('.btn-small.add').trigger('click');
    expect(wrapper.emitted('add-to-cart')[0]).toEqual([props.id]);
  });

  test('should render correctly', () => {
    const wrapper = createWrapper(ProductItem);
    expect(wrapper.element).toMatchSnapshot();
  });
});
