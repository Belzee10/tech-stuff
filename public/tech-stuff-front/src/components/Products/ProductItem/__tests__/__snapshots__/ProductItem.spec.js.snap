// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItem.vue should render correctly 1`] = `
<div
  class="product-item v-card v-card--link v-sheet theme--light"
  tabindex="0"
>
  <div
    class="container py-0"
  >
    <div
      class="row justify-space-between"
    >
      <div
        class="pa-0 col col-auto"
      >
        <div
          class="v-responsive v-image"
          style="height: 100px; width: 100px;"
        >
          <div
            class="v-image__image v-image__image--preload v-image__image--cover"
            mode="in-out"
            name="fade-transition"
            style="background-image: url(https://picsum.photos/id/11/10/6); background-position: center center;"
          />
          <div
            class="v-responsive__content"
          />
        </div>
      </div>
       
      <div
        class="pa-0 col col-5"
      >
        <div
          class="py-2 d-flex flex-column fill-height"
        >
          <h4
            class="body-1"
          >
            
          </h4>
           
          <span
            class="caption"
          >
            
          </span>
           
          <span
            class="font-weight-bold price d-block mt-auto"
          >
            $
          </span>
        </div>
      </div>
       
      <div
        class="text-center pa-0 col col-2"
      >
        <div
          class="row flex-column ma-0 fill-height justify-space-between"
        >
          <!---->
           
          <!---->
           
          <div
            class="pa-0 d-flex flex-column justify-center align-center col"
          >
            <button
              class="btn-full v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default secondary--text"
              type="button"
            >
              <span
                class="v-btn__content"
              >
                <i
                  aria-hidden="true"
                  class="v-icon notranslate mdi mdi-cart-plus theme--light"
                  style="font-size: 16px;"
                />
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
